pyincore/utils/datasetutil.py

```diff
@@ -63,3 +68,85 @@ def join_table_dataset_with_source_dataset(dataset, client):
+
+            # prepare retrofit definition into pandas dataframe; need to work with retrofit
+            if len(mapping.mappingEntryKeys) > 0:
+                mapping_entry_keys_df = pd.DataFrame(mapping.mappingEntryKeys)
+                # add suffix to avoid conflict
+                mapping_entry_keys_df.columns = [col + '_mappingEntryKey' for col in mapping_entry_keys_df.columns]
+                mapping_entry_keys_df.set_index('name_mappingEntryKey', inplace=True)
+                inventory_df = pd.merge(inventory_df, mapping_entry_keys_df, left_on='retrofit_key', right_index=True,
+                                        how='inner')
+                inventory_df.drop(columns=['defaultKey_mappingEntryKey'], inplace=True)
+            else:
+                raise ValueError("Missing proper definition for mappingEntryKeys in the mapping!")
+
+            def _apply_retrofit_value(row):
+                target_column = row["config_mappingEntryKey"]["targetColumn"] \
+                    if ("config_mappingEntryKey" in row.index and
+                        isinstance(row["config_mappingEntryKey"], dict) and
+                        "targetColumn" in row["config_mappingEntryKey"].keys()) else None
+                expression = row["config_mappingEntryKey"]["expression"] \
+                    if ("config_mappingEntryKey" in row.index and
+                        isinstance(row["config_mappingEntryKey"], dict) and
+                        "expression" in row["config_mappingEntryKey"].keys()) else None
+                type = row["config_mappingEntryKey"]["type"] \
+                    if ("config_mappingEntryKey" in row.index and
+                        isinstance(row["config_mappingEntryKey"], dict) and
+                        "type" in row["config_mappingEntryKey"].keys()) else None
+
+                if target_column and expression:
+                    if target_column in row.index:
+                        retrofit_value = float(row["retrofit_value"]) if type == "number" else row["retrofit_value"]
+
+                        # Dangerous! Be careful with the expression
+                        row[target_column] = eval(f"row[target_column]{expression}")
+                    else:
+                        raise ValueError(f"targetColumn: {target_column} not found in inventory properties!")
+
+                return row
+
+            inventory_df = inventory_df.apply(_apply_retrofit_value, axis=1)
+
+            # rename columns to fit the character limit of shapefile
+            inventory_df.rename(columns={
+                'retrofit_key': 'retrofit_k',
+                'retrofit_value': 'retrofit_v',
+                'description_mappingEntryKey': 'descr_map',
+                'config_mappingEntryKey': 'config_map'
+            }, inplace=True)
+
+            # save the updated inventory to a new shapefile
+            file_path = f"tmp_updated_{inventory_dataset.id}.shp"
+            inventory_df.to_file(file_path)
```

Comment:
Can I use system temp folder on k8s datawolf? Use random string for the temp folder name